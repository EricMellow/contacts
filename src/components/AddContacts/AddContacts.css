.add-contacts {
  margin: 0;
  padding: 20px;
  display: flex;
  justify-content: center;
  text-align: center;
}

.form-container {
  width: 100vw;
  height: 80vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-evenly;
}

.input-container {
  display: flex;
  flex-direction: column;
}

.inputs {
  font-size: 1rem;
  height: 2.5rem;
  margin: 10px;
  width: 350px;
}

.save-btn {
  font-size: 1.5rem;
  height: auto;
  width: auto;
  padding: 10px;
  border-radius: 4px;
  border-style: none;
  background-color: #106fb3;
  color: white;
}

.save-btn:hover {
  cursor: pointer;
}

small {
  text-align: start;
  margin-left: 10px;
}
