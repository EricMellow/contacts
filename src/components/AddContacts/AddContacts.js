import React, { Component } from "react";
import './AddContacts.css';
import { withRouter } from 'react-router-dom';
import { connect } from 'react-redux';
import { addContact } from '../../redux/actions';

class AddContacts extends Component {
  constructor() {
    super();
    this.state = {
      name: '',
      email: '',
      phone: '',
    };
  }

  handleChange = (event) => {
    this.setState({ [event.target.id]: event.target.value });
  }

  saveContact = (event) => {
    event.preventDefault();
    const contact = {
      id: Date.now(),
      name: this.state.name,
      email: this.state.email,
      phone: this.state.phone
    };
    this.props.addContact(contact);
    this.clearInputs();
  }

  clearInputs() {
    this.setState({
      name: '',
      email: '',
      phone: '',
    })
  }

  render() {
    return (
      <main>
        <form onSubmit={this.saveContact}>
          <label>Name: 
            <input
              type="text"
              id="name"
              name="name"
              value={this.state.name}
              onChange={this.handleChange}
              required />
          </label>
          
          <label>Email: 
            <input
              type="email"
              id="email"
              value={this.state.email}
              onChange={this.handleChange}
              required />
          </label>
          <small>Format: you@example.com</small>
          <label>Phone Number: 
            <input
              type="tel"
              id="phone"
              name="phone"
              pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}"
              value={this.state.phone}
              onChange={this.handleChange}
              required />
          </label>
          <small>Format: 123-456-7890</small>
          <input type="submit" value="Save Contact"></input>
        </form>
      </main>
    );
  }
}


export const mapDispatchToProps = (dispatch) => ({
  addContact: (contact) => dispatch(addContact(contact)),
});

export default withRouter(connect(null, mapDispatchToProps)(AddContacts));
