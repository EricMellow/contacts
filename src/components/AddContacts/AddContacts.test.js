import React from 'react';
import { AddContacts, mapDispatchToProps } from './AddContacts';
import { shallow } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

describe('AddContacts', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<AddContacts />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('clearInputs', () => {
    it('should setState with all empty strings', () => {
      const mockState = {
        name: 'Test Name',
        email: 'test@email.com',
        phone: '123-456-7890'
      };
      wrapper.setState(mockState);
      expect(wrapper.state('name')).toEqual('Test Name');
      expect(wrapper.state('email')).toEqual('test@email.com');
      expect(wrapper.state('phone')).toEqual('123-456-7890');
      wrapper.instance().clearInputs();
      expect(wrapper.state('name')).toEqual('');
      expect(wrapper.state('email')).toEqual('');
      expect(wrapper.state('phone')).toEqual('');
    })
  })

  describe('handleChange', () => {
    it('should set state with the id and value of the event passed to it', () => {
      const mockEmailEvent = {
        target: {
          id: 'email',
          value: 'test@email.com'
        }
      };
      expect(wrapper.state('email')).toEqual('');
      wrapper.instance().handleChange(mockEmailEvent);
      expect(wrapper.state('email')).toEqual('test@email.com');
    })

    it('should not change the change any other state values', () => {
      const mockEmailEvent = {
        target: {
          id: 'email',
          value: 'test@email.com'
        }
      };
      expect(wrapper.state('name')).toEqual('');
      expect(wrapper.state('phone')).toEqual('');
      wrapper.instance().handleChange(mockEmailEvent);
      expect(wrapper.state('name')).toEqual('');
      expect(wrapper.state('phone')).toEqual('');
    })

    it ('should work for every key-value pair in state', () => {
      const mockNameEvent = {
        target: {
          id: 'name',
          value: 'Test Name'
        }
      };

      const mockPhoneEvent = {
        target: {
          id: 'phone',
          value: '123-456-7890'
        }
      };
      expect(wrapper.state('name')).toEqual('');
      expect(wrapper.state('phone')).toEqual('');
      wrapper.instance().handleChange(mockNameEvent);
      expect(wrapper.state('name')).toEqual('Test Name');
      wrapper.instance().handleChange(mockPhoneEvent);
      expect(wrapper.state('phone')).toEqual('123-456-7890');
    })
  })

  describe('saveContact', () => {
    it('should call addContact with a contact object', () => {
      const mockState = {
        name: 'Test Name',
        email: 'test@email.com',
        phone: '123-456-7890'
      };
      const mockEvent = {
        preventDefault: jest.fn()
      }
      const mockAddContact = jest.fn();
      const wrapper = shallow(<AddContacts addContact={mockAddContact} />);
      wrapper.setState(mockState);
      wrapper.instance().saveContact(mockEvent);
      expect(wrapper.instance().props.addContact).toHaveBeenCalledWith({
        id: Date.now(),
        name: 'Test Name',
        email: 'test@email.com',
        phone: '123-456-7890'
      });

    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with the correct params on addContact', () => {
      const mockContact = {
        id: 402,
        name: 'Test Name',
        email: 'test@email.com',
        phone: '123-456-7890'
      };
      const mockDispatch = jest.fn();
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'ADD_CONTACT',
        contact: mockContact
      };

      mappedProps.addContact(mockContact);

      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });
  })

});
