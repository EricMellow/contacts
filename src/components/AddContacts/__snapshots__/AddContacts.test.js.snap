// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddContacts should match the snapshot 1`] = `ShallowWrapper {}`;
